### PR TITLE
IGNITE-23318 Add index/term/configuration to meta-storage

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-23318

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/SnapshotStorageFactory.java [hunk 1]
```diff
@@ -28,7 +28,7 @@ public interface SnapshotStorageFactory {
      *
      * @param uri Snapshot URI.
      * @param raftOptions Raft options.
-     * @return Snapshot storage.
+     * @return Snapshot storage. {@code null} if there's no snapshot.
      */
     SnapshotStorage createSnapshotStorage(String uri, RaftOptions raftOptions);
```
Comment (sashapolo) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/SnapshotStorageFactory.java:33]:
```suggestion
    @Nullable SnapshotStorage createSnapshotStorage(String uri, RaftOptions raftOptions);
```

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 2]
```diff
@@ -1659,6 +1751,25 @@ public long checksum(long revision) {
         }
     }
 
+    @Override
+    public void clear() {
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1755]: Do we have any tests for this method?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.List;
+import java.util.Objects;
+import org.apache.ignite.internal.tostring.IgniteToStringInclude;
+import org.apache.ignite.internal.tostring.S;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A RAFT group configuration at which it was committed.
+ */
+public class RaftGroupConfiguration implements Serializable {
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java:31]: I'm pretty sure I've seen an identical class somewhere...

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/MetaStorageSnapshotStorageFactory.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            @Override
+            public SnapshotReader open() {
+                if (startupSnapshotOpened.compareAndSet(false, true)) {
+                    if (startupSnapshotMeta == null) {
+                        // The storage is empty, let's behave how JRaft does: return null, avoiding an attempt to load a snapshot
+                        // when it's not there.
+                        return null;
+                    }
+
+                    return new StartupMetaStorageSnapshotReader(MetaStorageSnapshotStorageFactory.this);
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/MetaStorageSnapshotStorageFactory.java:102]: This is so weird. Why don't we pass the `startupSnapshotMeta` here and instead have this circular reference?

Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/MetaStorageSnapshotStorageFactory.java:102]: Even more, I would expect the `StartupMetaStorageSnapshotReader` to read the startup meta and instead it is done by the factory for some reason

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 5]
```diff
@@ -220,6 +229,14 @@ public void onSnapshotSave(Path path, Consumer<Throwable> doneClo) {
 
     @Override
     public boolean onSnapshotLoad(Path path) {
+        // Startup snapshot should always be ignored.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:232]: Expanded the comment

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java (hunk 2 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1755]: No, I'll add some

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java (hunk 3 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java:31]: Yes, that's a copy-paste

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/MetaStorageSnapshotStorageFactory.java (hunk 4 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/MetaStorageSnapshotStorageFactory.java:102]: >  read the startup meta and instead it is done by the factory for some reason

I have to read meta in order to determine whether I need to return `null` or `StartupMetaStorageSnapshotReader`. When it comes to circular references, my bad

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java (hunk 3 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java:31]: Shall we address this somehow?

File: modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageServiceTest.java [hunk 6]
```diff
@@ -318,10 +317,8 @@ private List<Node> prepareNodes(int amount) {
     }
 
     /**
-     * Starts nodes. It is important to call {@link #prepareNodes(int)}, then configure required mocks, and only then star nodes. Otherwise
-     * the {@link Node#mockStorage} (for example) will be concurrently used by a node and configured in the test (using a
-     * {@link Mockito#when(Object)}, for example). This is not allowed, mocks must be accessed exclusively during a configuration phase,
-     * it's not thread-safe.
+     * Starts nodes. It is important that this method is called before all mocks are configured, otherwise we will have races between raft
```
Comment (sashapolo) [inline modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageServiceTest.java:320]: > It is important that this method is called before all mocks are configured

Are you sure?

Comment (ibessonov) [reply modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageServiceTest.java:320]: No

**Review by sashapolo — APPROVED** [review]

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java (hunk 3 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/raft/RaftGroupConfiguration.java]: Yes, I removed the copy-paste and moved this class to `raft-api`

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 7]
```diff
@@ -207,7 +209,14 @@ public boolean onBeforeApply(Command command) {
 
     @Override
     public void onConfigurationCommitted(CommittedConfiguration config) {
-        RaftGroupListener.super.onConfigurationCommitted(config);
+        RaftGroupConfiguration configuration = new RaftGroupConfiguration(
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:212]: Yes, we can

File: modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngineTest.java [hunk 8]
```diff
@@ -426,7 +427,7 @@ void replicasTriggersAssignmentsChangingOnNonDefaultZones() throws Exception {
 
         keyValueStorage.put(
                 stablePartAssignmentsKey(new TablePartitionId(getTableId(TABLE_NAME), 0)).bytes(), assignmentsBytes,
-                clock.now()
+                new KeyValueUpdateContext(0, 0, clock.now())
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngineTest.java:430]: No, that method is not visible in this class, it's not in the hierarchy

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java [hunk 9]
```diff
@@ -225,12 +225,12 @@ void testSaveCompactionRevisionDoesNotChangeRevisionInMemory() {
     }
 
     @Test
-    void testSaveCompactionRevisionAndRestart() throws Exception {
+    public void testSaveCompactionRevisionAndRestart() throws Exception {
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:228]: Can make it protected if you wish. I override it in subclass

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryCompactionKeyValueStorageTest.java [hunk 10]
```diff
@@ -23,4 +23,9 @@ public class SimpleInMemoryCompactionKeyValueStorageTest extends AbstractCompact
     public KeyValueStorage createStorage() {
         return new SimpleInMemoryKeyValueStorage("test");
     }
+
+    @Override
+    public void testSaveCompactionRevisionAndRestart() {
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryCompactionKeyValueStorageTest.java:28]: Same answer. Should it be protected?

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorageTest.java [hunk 11]
```diff
@@ -42,6 +43,11 @@ public void tearDown() throws Exception {
         storage.close();
     }
 
+    /** Creates a Meta Storage context based on a timestamp. Index and term will be {@code 0}. */
+    protected static KeyValueUpdateContext msContext(HybridTimestamp timestamp) {
+        return new KeyValueUpdateContext(0, 0, timestamp);
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorageTest.java:48]: I'm not a big fan of creating test-only methods in production classes. What's the benefit?

[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorageTest.java:48]: > We already have such test-only method with awkward placement and using it half the time

That's a bad justification. We have a lot of bad code, should I write bad code too? :)

[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorageTest.java:48]: Anyway, you won, I'm too lazy today

**Pull request merged**